mindmeld/models/nn_utils/classification.py

```diff
@@ -0,0 +1,490 @@
+        self.params["name"] = self.name
+        self.encoder = None
+
+        self.ready = False  # True when .fit() is called or loaded from a checkpoint, else False
+        self.dirty = False  # True when the model weights aren't saved to disk yet, else False
+
+    def __repr__(self):
+        return f"<{self.name}> ready:{self.ready} dirty:{self.dirty}"
+
+    def get_default_params(self) -> Dict:
+        return get_default_params(self.__class__.__name__)
+
+    def who_am_i(self) -> str:
+        msg = f"{self.name} " \
+              f"ready:{self.ready} dirty:{self.dirty} device:{self.params.device}\n" \
+              f"\tNumber of weights (trainable, all):{get_num_weights_of_model(self)} \n" \
+              f"\tDisk Size (in MB): {get_disk_space_of_model(self):.4f}"
+        logger.info(msg)
+        return msg
+
+    def _to_device(self, batch_data: BatchData) -> BatchData:
+        """method for gpu usage and data loading"""
+        for k, v in batch_data.items():
+            if v is not None and isinstance(v, torch.Tensor):
+                batch_data[k] = v.to(self.params.device)
+            elif isinstance(v, list):
+                batch_data[k] = [
+                    vv.to(self.params.device) if isinstance(vv, torch.Tensor) else vv
+                    for vv in v
+                ]
+            elif isinstance(v, dict):
+                batch_data[k] = self._to_device(batch_data[k])
+        return batch_data
+
+    def fit(self, examples, labels, **params):  # pylint: disable=too-many-locals
+        """
+        Trains the underlying neural model on the inputted data and finally retains the best scored
+        model among all iterations.
+
+        Because of possibly large sized neural models, instead of retaining a copy of best set of
+        model weights on RAM, it is advisable to dump them in a temporary folder and upon training,
+        load the best checkpointed weights.
+        """
+
+        if self.ready:
+            msg = "The model is already fitted or is loaded from a file. Aborting re-fitting."
+            logger.error(msg)
+
+        # update params upon preparing encoder and embedder
+        params = {**self.get_default_params(), **params}
+        params = self._prepare_input_encoder(examples, **params)
+        params = self._prepare_embedder(**params)
+
+        # update params upon identifying unique labels
+        try:
+            num_labels = len(set(labels))
+        except TypeError:  # raised in cased on token classification
+            num_labels = len(set(sum(labels, [])))
+        params.update({"num_labels": num_labels})
+
+        # update params with label pad idx
+        try:  # used by token classifiers in crf masks and for training/evaluation
+            self.label_padding_idx = params["label_padding_idx"]
+            params.update({"label_padding_idx": params["label_padding_idx"]})
+        except KeyError:  # raised in cased on sequence classification
```

Comment:
Nit: Can you elaborate on what this exception is exactly (what does missing label_padding_idx signify maybe?) and which fail gets it here? Not sure if the comment is a typo but not completely clear!

Reply:
so in case of token classification, for the labels, we pass a label pad id which is passed to cross entropy loss. However, for seq classification (i.e sentence level text classification), no such idx is required and hence we wont set it. 

Because we have a single fit() method irrespective of the seq or token classification, we are just checking if that pad idx is available or not and if not, we just ignore.

I'll modify the comment to make it clear.